File: frontend/src/types/Sprint.ts
export type SprintStatus =
  | 'PLANNED'
  | 'ACTIVE'
  | 'IN_PROGRESS'
  | 'READY_FOR_APPROVAL'
  | 'COMPLETED'
  | 'ARCHIVED';

export interface Sprint {
  id: number;
  name: string;
  goal: string | null;
  start_date: string;
  end_date: string;
  actual_start_date: string | null;
  completed_at: string | null;
  estimated_team_members: number | null;
  working_days: number | null;
  hours_per_day: number | null;
  goal_status: string;
  status: SprintStatus;
  project_id: number;
  created_at: string;
  updated_at: string;
  // Approval workflow
  assigned_tester_id: number | null;
  assigned_tester_name: string | null;
  submitted_by_id: number | null;
  submitted_at: string | null;
  approved_by_id: number | null;
  approved_at: string | null;
  review_comment: string | null;
}

export interface SprintCreate {
  name: string;
  goal?: string | null;
  start_date: string;
  end_date: string;
  project_id: number;
  estimated_team_members?: number | null;
  working_days?: number | null;
  hours_per_day?: number | null;
  issue_ids?: number[];
}

export interface SprintUpdate {
  name?: string;
  goal?: string | null;
  start_date?: string;
  end_date?: string;
  status?: SprintStatus;
  estimated_team_members?: number | null;
  working_days?: number | null;
  hours_per_day?: number | null;
  goal_status?: string | null;
}

export interface SprintExtend {
  new_end_date: string;
}

export interface SprintAssignTester {
  tester_id: number;
}

export interface SprintRequestChanges {
  comment?: string | null;
}

export interface SprintOverview {
  total_sprints: number;
  active_sprint: Sprint | null;
  completed_sprints: number;
  avg_completion_rate: number;
  avg_velocity: number;
  overdue_sprints: number;
}

export interface SprintAnalytics {
  total_issues: number;
  completed_issues: number;
  remaining_issues: number;
  completion_rate: number;
  open_issues: number;
  in_progress_issues: number;
  resolved_issues: number;
  closed_issues: number;
  total_capacity_hours: number | null;
  workload: {
    developer_id: number;
    developer_name: string;
    assigned_issues: number;
    completed_issues: number;
    in_progress_issues: number;
    open_issues: number;
  }[];
  burndown_points: {
    date: string;
    remaining: number;
    ideal: number;
  }[];
  sprint_health: 'ON_TRACK' | 'AT_RISK' | 'OFF_TRACK' | null;
  is_overdue: boolean;
  days_overdue: number;
  total_estimated_effort: number;
  completed_effort: number;
  remaining_effort: number;
}

File: backend/app/schemas/sprint.py
from datetime import datetime
from pydantic import BaseModel, ConfigDict, Field
from app.models.sprint import SprintStatus


class SprintBase(BaseModel):
    name: str = Field(..., max_length=200, example="Sprint 1")
    goal: str | None = Field(None, example="Finish login flow")
    start_date: datetime
    end_date: datetime
    estimated_team_members: int | None = Field(None, ge=1)
    working_days: int | None = Field(None, ge=1)
    hours_per_day: int | None = Field(None, ge=1)
    goal_status: str = Field("NOT_STARTED", max_length=50)


class SprintCreate(SprintBase):
    project_id: int
    issue_ids: list[int] | None = Field(None, description="Issues to assign to sprint on creation")


class SprintUpdate(BaseModel):
    name: str | None = Field(None, max_length=200)
    goal: str | None = None
    start_date: datetime | None = None
    end_date: datetime | None = None
    status: SprintStatus | None = None
    estimated_team_members: int | None = None
    working_days: int | None = None
    hours_per_day: int | None = None
    goal_status: str | None = Field(None, max_length=50)

class SprintExtend(BaseModel):
    new_end_date: datetime


# ── Approval workflow request schemas ─────────────────────────────────────────

class SprintAssignTester(BaseModel):
    tester_id: int = Field(..., description="ID of the TESTER user to assign")


class SprintRequestChanges(BaseModel):
    comment: str | None = Field(None, max_length=2000, description="Optional reason/feedback for the tester")


# ── Read schema ───────────────────────────────────────────────────────────────

class SprintRead(SprintBase):
    id: int
    project_id: int
    status: SprintStatus
    actual_start_date: datetime | None
    completed_at: datetime | None
    created_at: datetime
    updated_at: datetime
    # Approval workflow fields
    assigned_tester_id: int | None = None
    assigned_tester_name: str | None = None   # resolved from relationship
    submitted_by_id: int | None = None
    submitted_at: datetime | None = None
    approved_by_id: int | None = None
    approved_at: datetime | None = None
    review_comment: str | None = None

    model_config = ConfigDict(from_attributes=True)

    @classmethod
    def model_validate(cls, obj, **kwargs):
        # Resolve related user names if ORM object without triggering lazy IO
        instance = super().model_validate(obj, **kwargs)
        try:
            if hasattr(obj, "__dict__"):
                tester = obj.__dict__.get("assigned_tester")
                if tester and hasattr(tester, "full_name"):
                    instance.assigned_tester_name = tester.full_name
        except Exception:
            pass
        return instance


class SprintAnalytics(BaseModel):
    total_issues: int
    completed_issues: int
    remaining_issues: int
    completion_rate: float
    open_issues: int
    in_progress_issues: int
    resolved_issues: int
    closed_issues: int
    total_capacity_hours: int | None
    workload: list[dict]
    burndown_points: list[dict]
    sprint_health: str | None = None
    is_overdue: bool = False
    days_overdue: int = 0
    total_estimated_effort: int = 0
    completed_effort: int = 0
    remaining_effort: int = 0

class SprintOverview(BaseModel):
    total_sprints: int
    active_sprint: SprintRead | None
    completed_sprints: int
    avg_completion_rate: float
    avg_velocity: float
    overdue_sprints: int

